Make Python3 friendly and support category facet relevance

rlvoyer opened the pull request:
This changeset includes modifications to make arcs work in Python3. Additionally, and more importantly, it adds support for the category/search facet relevance jobs (specifically, to allow us to compare relevance on category facet queries w/ old and new clads).

Commit:
Make Python3 friendly

Commit:
Accommodate category facet relevance task

Commit:
Add num_rows and num_columns args to launch_jobs script

Commit:
Use right default cetera host and handle failures when uploading to CrowdFlower

Commit:
Fix job results fetching from CrowdFlower

Commit:
Remove cruft

Commit:
Add missing files, improve .gitignore, and handle requests exceptions

kavedder commented on arcs/cetera.py:
```diff
@@ -0,0 +1,120 @@
+        cetera_host (str): The Cetera hostname
+        cetera_port (int): The port on which to make requests to Cetera
+        num_results (int): The number of results to fetch for each query
+        cetera_params (Dict[Any, Any]): Optional, additional query parameters for Cetera
```
such scala ;-)

rlvoyer replied:
It appears that way, but did you check out Python 3 type annotations? As I mentioned previously, there is no comprehensive guide on the right way to do this, so I'm synthesizing the Google Python styleguide and the docs on the Python 3 type annotations (https://docs.python.org/3/library/typing.html). I find it awesomely readable.

kavedder commented on arcs/crowdflower.py:
```diff
@@ -34,7 +34,8 @@ def job_from_dict(job_data):
     return Job(**data)
 
-headers = frozendict({'content-type': 'application/json', 'accept': 'application/json'})
+headers = frozendict({'content-type': 'application/json; charset=utf-8',
```
Did 2to3 do this for you? If so, impressive!

rlvoyer replied:
No. I did that.

kavedder commented on arcs/crowdflower.py:
```diff
@@ -46,11 +47,13 @@ def create_job_from_copy(api_key=None, job_id=None):
+        A new Job
     """
-    job_id = job_id or 844991
+    job_id = job_id or 911946
```
Do we really want to hardcode this?

rlvoyer replied:
Not at all. I thought it would be useful, but ultimately, it was a terrible idea, because it makes it easy to forget to update this default, which you want to do to get the most current set of test data for a task.

kavedder commented on arcs/crowdflower.py:
```diff
@@ -46,11 +47,13 @@ def create_job_from_copy(api_key=None, job_id=None):
+    job_id = job_id or 911946
     api_key = api_key or os.environ["CROWDFLOWER_API_KEY"]
-    url = "https://api.crowdflower.com/v1/jobs/{}/copy.json?key={}&gold=true".format(job_id, api_key)
+    url = "https://api.crowdflower.com/v1/jobs/{}/copy.json?key={}&gold=true".format(
```
Generally I feel like we should be passing params as a `params` dict, but this is not necessary for this commit.

rlvoyer replied:
You are 100% right.

kavedder commented on arcs/crowdsourcing_utils.py:
```diff
@@ -0,0 +1,336 @@
+
+def cleanup_description(desc, nlp=None):
+    """
+    Make a dataset description is readable and not ridiculously long.
```
Nit: `Make a dataset description is readable` -> `Make sure a dataset description is readable`? `Make a dataset description readable`?

kavedder commented on arcs/crowdsourcing_utils.py:
```diff
@@ -0,0 +1,336 @@
+    analyzer = nlp or NLP
+    desc = desc.replace("\r", "\n")
+    desc_doc = analyzer(desc) if desc else desc
+    desc_sentences = [s.text.replace("\n", " ").strip() for s in desc_doc.sents] if desc else []
```
Should probably actually be `re.sub(r'\s+', ' ', s.text).strip()`. There's a lot of `\r\n` in the descriptions we get.

kavedder commented on arcs/crowdsourcing_utils.py:
```diff
@@ -0,0 +1,336 @@
+    return _join_sentences("", desc_sentences, 400) if desc else desc
+
+
+def extract_address(cell_contents):
```
So much duplicate code...putting this in a shared library is going to be awesome.

rlvoyer replied:
Word. I'm excited for that.

kavedder commented on arcs/launch_job.py:
```diff
@@ -1,193 +1,115 @@
 
     Args:
-        desc (str): A dataset (or other core object) description
+        result ():
```
Merp?

rlvoyer replied:
Derp.

kavedder commented on setup.py:
```diff
@@ -9,18 +9,16 @@ def read(fname):
-                         'requests[security]==2.7.0',
+install_requires_list = ['pandas>=0.18.1',
+                         'matplotlib>=1.5',
+                         'numpy>=1.11.0',
```
Isn't installing numpy thusly somehow borked?

rlvoyer replied:
I don't think so. What makes you say that?

kavedder commented on task_templates/crowdflower_instructions.html:
```diff
@@ -0,0 +1,168 @@
+<div class="row-fluid">
```
While you're checking in html files, might be good to check in the CSS as well.

kavedder commented:
As with any large commit, there are probably things both of us missed, but they'll come out as we use this. It would be really nice to move all the code shared between here and clads into a separate project :-) it's a little painful seeing it copied into here, but this is good for now.

Commit:
Cleanup crowdflower and crowdsourcing_utils modules per PR comments

- made job_id required parameter for crowdflower `create_job_from_copy`
- used requests params rather than string interpolation for query params
- fixed docstring
- add task custom CSS to repo
